### PR TITLE
Add Inspector tab

### PR DESCRIPTION
**Pull request by freeall**
Adds inspect feature. Works together with https://github.com/holepunchto/pear/pull/411.

image

How to test:

1. Use https://github.com/holepunchto/pear/pull/411
2. Create Pear terminal app (`{"pear": {"type": "terminal"}}` in the app's `package.json`)
3. Run the app like `$ pear.dev dev . --inspect`. The app will output a 64-string long key
4. Paste the key into the field in the Inspector tab in Pear Pulse
5. Open `chrome://inspect` and see your app there

### CODE REVIEW

File: lib/devtools-inspector.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+          </form>
+        </div>
+        <h3 id="no-apps">No apps availble. Add a key above, to start debugging.</h3>
+        <div id="apps"></div>
+      </div>
+    `
+    this.root = this.attachShadow({ mode: 'open' })
+    this.root.appendChild(this.template.content.cloneNode(true))
+
+    this.$addKeyForm = this.root.querySelector('#add-key-form')
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: Seems like we are not using the $name convention, see here https://github.com/holepunchto/pear-pulse/blob/main/lib/docs-viewer.js#L55

File: lib/devtools-inspector.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    this.$addKey = this.root.querySelector('#add-key')
+    this.$addKeyError = this.root.querySelector('#add-key-error')
+    this.$apps = this.root.querySelector('#apps')
+    this.$noApps = this.root.querySelector('#no-apps')
+    this.apps = new Map()
+
+    this.$addKeyForm.addEventListener('submit', e => {
+      e.preventDefault()
+      const key = this.$addKey.value
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: the error reset should happen on a 'change' event maybe?

File: lib/devtools-inspector.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+      const targets = [...this.apps].map(([sessionId]) => ({
+        description: 'node.js instance', // `Pear app: ${app.name}`,
+        devtoolsFrontendUrl: `devtools://devtools/bundled/js_app.html?experiments=true&v8only=true&ws=127.0.0.1:9229/${sessionId}`,
+        devtoolsFrontendUrlCompat: `devtools://devtools/bundled/inspector.html?experiments=true&v8only=true&ws=127.0.0.1:9229/${sessionId}`,
+        faviconUrl: 'https://nodejs.org/static/images/favicons/favicon.ico',
+        id: sessionId,
+        title: 'index.js',
+        type: 'node',
+        url: `file:///path/to/some/file/index.js`,
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: is this a placeholder?

File: lib/devtools-inspector.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      res.writeHead(200, {
+        'Content-Type': 'application/json; charset=UTF-8',
+        'Cache-Control': 'no-cache',
+        'Content-Length': JSON.stringify(targets).length
+      })
+      res.end(JSON.stringify(targets))
+    })
+    httpServer.on('upgrade', (request, socket, head) => {
+      console.log(`[httpServer] UPGRADE. url=${request.url}`)
+      const sessionId = request.url.substr(1)
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: does this remove '/'? is there any chance that its not part of the url?

File: lib/devtools-inspector.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        devtoolsSocket.send(JSON.stringify(msg))
+      })
+      devtoolsSocket.on('message', msg => {
+        inspectorClient.send(JSON.parse(msg))
+      })
+    })
+  }
+
+  renderApps () {
+    const content = [...this.apps.values()].map(app => `
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: maybe `this.apps.values().map(app => ....)` works too?

Comment (freeall) [reply lib/devtools-inspector.js]: Just tried it out, and apparently nope :)

File: lib/devtools-inspector.js (hunk 4 above)
Comment (freeall) [reply lib/devtools-inspector.js]: Yeah, it just removes that `/`. There shouldn't be any chance that it's not part of the url as it's part of how `chrome://inspect` works.

File: lib/devtools-inspector.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        this.addKeyErrorElem.textContent = ''
+        return
+      }
+      if (publicKey.length !== 64) {
+        this.addKeyErrorElem.textContent = 'Key needs to be 64 characters long'
+        return
+      }
+
+      const sessionId = generateUuid()
+      const inspectorSession = new Session({ publicKey: Buffer.from(publicKey, 'hex') })
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: I think `b4a` its prefered over `Buffer`

File: lib/devtools-inspector.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    this.addKeyErrorElem = this.root.querySelector('#add-key-error')
+    this.appsElem = this.root.querySelector('#apps')
+    this.noAppsElem = this.root.querySelector('#no-apps')
+    this.apps = new Map()
+
+    this.addKeyFormElem.addEventListener('submit', e => {
+      e.preventDefault()
+      const publicKey = this.addKeyInputElem.value
+      if (publicKey.length === 0) {
+        this.addKeyErrorElem.textContent = ''
```
Comment (rafapaezbas) [inline lib/devtools-inspector.js]: should the error reset happen on 'change' instead of submit?

File: package.json [hunk 8]
```diff
@@ -10,7 +10,13 @@
         }
     },
     "dependencies": {
+        "@holepunchto/pear-inspect": "^3.0.1",
+        "b4a": "^1.6.4",
         "bare-path": "^2.1.0",
-        "overpass": "github:RedHatOfficial/Overpass#71f18db"
+        "overpass": "github:RedHatOfficial/Overpass#71f18db",
```
Comment (davidmarkclements) [inline package.json]: please use https://www.npmjs.com/package/redhat-overpass-font

**Comment by davidmarkclements** [comment]
landing screen bug 
image

**Review by davidmarkclements — CHANGES_REQUESTED** [review]
typo: availble, landing screen bug, replace gh package font with npm package, call the section "Developer Tooling" instead of "Inspector" (inspector is a subsection which happens to be the first, this section will grow as we go)

**Pull request merged**